src/datasets/arrow_writer.py

```diff
@@ -428,12 +428,17 @@ def write_examples_on_file(self):
             return
 
         # order the columns properly
-        cols = (
-            [col for col in self.schema.names if col in self.current_examples[0][0]]
-            + [col for col in self.current_examples[0][0].keys() if col not in self.schema.names]
-            if self.schema
-            else self.current_examples[0][0].keys()
-        )
+        if self.schema:
+            schema_names_set = set(self.schema.names)
+            current_example_keys_set = set(self.current_examples[0][0].keys())
+
+            common_cols = schema_names_set & current_example_keys_set
+            extra_cols = current_example_keys_set - schema_names_set
+
+            cols = list(common_cols) + list(extra_cols)
+        else:
+            cols = self.current_examples[0][0].keys()
```

Comment:
Python sets (unlike dict keys) don't preserve the order, so we need to use lists to construct `common_cols` and `extra_cols`.

```suggestion
        if self.schema:
            schema_cols = set(self.schema.names)
            common_cols, extra_cols = [], []
            for col in self.current_examples[0][0]:
                if col in schema_cols:
                    common_cols.append(col)
                else:
                    extra_cols.append(col) 
            cols = common_cols + extra_cols
        else:
            cols = list(self.current_examples[0][0])
```